### DISCUSSION TITLE
Symbols panels show markdown headers are nested

### DISCUSSION THREAD

**Discussion by lunarwhite** [community/community#82803] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Bug

### Body

Take https://github.com/community/community/blob/main/CODE_OF_CONDUCT.md?plain=1 for an example.

Open `Symbols panels`, we will see every markdown headers are nested, no matter H1/H2/H3.

```
## Overview and Purpose

## Pledge

## Standards

### _Best Practices for Maintaining a Strong Community_

### _What is not Allowed_

...

```

It's weird that symbols didn't keep their original structures.

![Screenshot 2023-12-28 21 34 23](https://github.com/community/community/assets/57584831/cfbee869-a5f7-4fb4-a41b-10f2ca9d9408)

**Comment by muzimuzhi** [comment]
It's still the case in Sep 2024. See for example https://github.com/actions/checkout/blob/6d193bf28034eafb982f37bd894289fe649468fc/CHANGELOG.md?plain=1

<img width="1043" alt="image" src="https://github.com/user-attachments/assets/1f0038ea-86d6-4609-a6e6-00b7853f4bd0">

**Reply by muzimuzhi** [reply]
Checked on Jun 22, 2025 and the problem persists.

In the Preview tab, there's no `plain=1` URL parameter and markdown headers result in an Outline panel. All fine.

In the Code tab, `plain=1` parameter is added and markdown headers result in nested symbols in an Symbols panel.

**Comment by dannz510** [comment]
Excellent bug report. You've clearly identified an issue with how the "Symbols" or "Outline" view parses Markdown files. The panel should reflect the semantic hierarchy of the headers, but instead, it seems to be nesting every header under the one that immediately precedes it in the document.

Here is a more formally structured bug report based on your findings.

-----

### **Bug Report**

**Title:** Markdown "Symbols" Outline Incorrectly Nests Headers Instead of Maintaining Document Hierarchy

**Component:** File Viewer / Editor Symbols Panel

**Description:**
When viewing a Markdown file, the "Symbols" panel (which provides a document outline based on headers) does not correctly represent the file's structure. Instead of creating a hierarchy where headers of the same level are siblings and lower-level headers are children of the preceding higher-level header, the panel nests every header under the one directly before it.

This creates a deeply and incorrectly nested list, making the outline difficult to use for navigating documents with a standard structure.

**Steps to Reproduce:**

1.  Create or open a Markdown file with a standard header structure, such as the `CODE_OF_CONDUCT.md` from `github/community`.
    ```markdown
    ## Overview and Purpose
    Text...

    ## Pledge
    Text...

    ## Standards
    Text...

    ### Best Practices
    Text...

    ### What is not Allowed
    Text...

    ## Enforcement
    Text...
    ```
2.  Open the "Symbols" or "Outline" view for this file.

**Expected Behavior:**
The outline should treat headers of the same level as siblings. The structure should look like this, reflecting the actual document hierarchy:

```
- Overview and Purpose  (H2)
- Pledge              (H2)
- Standards           (H2)
  - Best Practices      (H3)
  - What is not Allowed (H3)
- Enforcement         (H2)
```

**Actual Behavior:**
The outline incorrectly nests each header under the previous one, regardless of its level, creating a confusing and inaccurate chain:

```
- Overview and Purpose  (H2)
  - Pledge              (H2)
    - Standards           (H2)
      - Best Practices      (H3)
        - What is not Allowed (H3)
          - Enforcement         (H2)  <-- This is especially wrong
```

**Conclusion:**
The logic for generating the symbol tree for Markdown appears to be based on linear order rather than heading level (`#`, `##`, `###`), which defeats the purpose of a hierarchical outline view.
